consensus/src/sender_aware_shuffler.rs

```diff
@@ -0,0 +1,464 @@
+use move_core_types::account_address::AccountAddress;
+use std::collections::{HashMap, VecDeque};
+
+/// An implementation of transaction shuffler, which tries to spread transactions from same senders
+/// in a block in order to reduce conflict. On a high level, it works as follows - It defines a
+/// `conflict_window_size`, which maintains a set of senders added to the block in last `conflict_window_size`
+/// transactions. When trying to select a new transaction to the block, the shuffler tries to find
+/// a transaction which are not part of the conflicting senders in the window. If it does, it adds
+/// the first non-conflicting transaction it finds to the block, if it doesn't then it preserves the
+/// order and adds the first transaction in the remaining block. It always maintains the following
+/// invariant in terms of ordering
+/// 1. Relative ordering of all transactions from the same before and after shuffling is same
+/// 2. Relative ordering of all transactions across different senders will also be maintained if they are
+/// non-conflicting. In other words, if the input block has only one transaction per sender, the output
+/// ordering will remain unchanged.
+///
+/// In order to make the look up faster, it creates an index of contiguous set of transactions by
+/// senders. This makes searching for non-conflicting transactions much faster as compared to brute
+/// force approach.
+///
+/// Another optimization introduced is the `look_ahead_sender_window` - this limits the number of senders
+/// to check to find non-conflicting transactions before giving up. This is needed to ensure the shuffling
+/// is not O(n*2) even in the worst case.
+
+pub struct SenderAwareShuffler {
+    conflict_window_size: usize,
+}
+
+impl TransactionShuffler for SenderAwareShuffler {
+    fn shuffle(&self, txns: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
+        let _timer = TXN_SHUFFLE_SECONDS.start_timer();
+
+        // maintains the intermediate state of the shuffled transactions
+        let mut sliding_window = SlidingWindowState::new(self.conflict_window_size);
+        let mut pending_txns = PendingTransactions::new();
+        let num_transactions = txns.len();
+        let mut orig_txns = VecDeque::from(txns);
+        'outer: loop {
+            if sliding_window.num_txns() == num_transactions {
+                break;
+            }
+            // First check if we have a sender dropped off of conflict window in previous step, if so,
+            // we try to find pending transaction from the corresponding sender and add it to the block.
+            if let Some(sender) = sliding_window.last_dropped_sender() {
+                if let Some(txn) = pending_txns.remove_pending_from_sender(sender) {
+                    sliding_window.add_transaction(txn);
+                    continue 'outer;
+                }
+            }
+
+            // Iterate through the original transactions and try to find the next candidate
```

Comment:
this should be in else? once we add another transaction, window moves, and we can have another sender that can be added back?

for example, window=3
input:
A1, A2, A3, B1, B2, B3, C1, D1

output should be 
A1, B1, C1, A2, B2, D1, A3, B3

I think your code will do

 A1, B1, C1, A2, D2 ...

Reply:
We have `continue` in the loop above, so having `else` will have the same effect.